File: ICS/Practicals/RSA.java
import java.math.BigInteger;
import java.util.Scanner;

/*
    STEPS:
    1) Choose two (p, q) different large prime number
    2) calculate n = p * q
    3) calculate phi(n) = (p - 1) * (q - 1)
    4) Choose e, 1 < e < phi(n) such that gcd(e, phi(n)) = 1
    5) calculate d = e^-1 mod phi(n)

    e,n = public key
    d,n = private key
*/

public class RSA {

    static long gcd(long phiOfN, long e) {
        long temp;
        while (true) {
            temp = phiOfN % e;
            if (temp == 0)
                return e;
            phiOfN = e;
            e = temp;
        }
    }

    public static long getEncyptionkey(long phiOfN) {
        long e = 2;
        for (e = 2; e < phiOfN; e++) {
            if (gcd(phiOfN, e) == 1) {
                return e;
            }
        }
        return e;
    }

    public static long getDecryptionKey(long phiOfN, long e) {
        double num = 0.0;
        for (int i = 0; i < 100; i++) {
            num = (double) (i * phiOfN + 1) / e;
            if (Math.floor(num) == num)
                break;
        }
        return (long) num;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter two prime number (p & q) ");
        int p = sc.nextInt();
        int q = sc.nextInt();

        long n = p * q;

        long phiOfN = (p - 1) * (q - 1);

        long e = getEncyptionkey(phiOfN);

        long d = getDecryptionKey(phiOfN, e);

        System.out.println("n = " + n);
        System.out.println("e (public key) = " + e);
        System.out.println("d (private key) = " + d);
        while (true) {
            System.out.println("Enter message to be encrypted ");
            long message = sc.nextLong();
            if (message == 0)
                break;
            BigInteger encryptedMessage = (BigInteger.valueOf(message).pow((int) e).mod(BigInteger.valueOf(n))); // msg
                                                                                                                 // ^ e
                                                                                                                 // mod
                                                                                                                 // n
            System.out.println("Encrypted message = " + encryptedMessage.intValue());

            BigInteger decryptedMessage = (BigInteger.valueOf(encryptedMessage.intValue()).pow((int) d)
                    .mod(BigInteger.valueOf(n))); // enc ^ d mod n
            System.out.println("Decrypted message = " + decryptedMessage.intValue());
        }
        sc.close();
    }
}

File: ICS/Practicals/RSA.py
"""
    STEPS:
    1) Choose two (p, q) different large prime number
    2) calculate n = p * q
    3) calculate phi(n) = (p - 1) * (q - 1)
    4) Choose e, 1 < e < phi(n) such that gcd(e, phi(n)) = 1
    5) calculate d = e^-1 mod phi(n)

    e,n = public key
    d,n = private key

"""
import math
import socket

SERVER_PORT = 1234


def gcd(a, b):
    while True:
        temp = a % b
        if temp == 0:
            return b
        a = b
        b = temp


def getEncryptionKey(phiOfN):
    for i in range(2, phiOfN):
        if gcd(phiOfN, i) == 1:
            return i


def getDecryptionKey(phiOfN, e):
    d = 0.0
    for i in range(1, 100):
        d = float(i*phiOfN+1)/e
        if math.floor(d) == d:
            print("i =", i, "d =", d)
            break
    return int(d)


def main():

    serverIP = input("Enter server's IP: (leave blank) ")
    if serverIP != "":
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect((serverIP, SERVER_PORT))

        data = sock.recv(1024)
        data = data.decode()
        print("received message:", data)
        n, e = map(int, data.split("|"))

        while True:
            msg = int(input("Enter msg to send (0 to quit) : "))
            if(msg == 0):
                break
            enc = pow(msg, e, n)
            print("encrypted message:", enc)
            sock.sendall(f"{enc}".encode())
        return

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind(('', SERVER_PORT))
    sock.listen(5)

    p = int(input("Enter prime no. (p): "))
    q = int(input("Enter prime no. (q): "))
    n = p * q
    phiOfN = (p - 1) * (q - 1)

    e = getEncryptionKey(phiOfN)
    d = getDecryptionKey(phiOfN, e)

    print(f"p = {p}\tq= {q}")
    print(f"n = {n}\tphi(n) = {phiOfN}")
    print(f"Public Key(e) = {e}")
    print(f"Private Key(d) = {d}")

    print("\n---Waiting for connection---")
    connection, client_address = sock.accept()

    print(f"\nSending public key to {client_address}")
    connection.sendall(f"{n}|{e}".encode())
    print("Public key sent")

    while True:
        data = connection.recv(1024)
        print("encrypted message:", data.decode())
        data = int(data.decode())
        print("decrypted message:", pow(data, d, n))

    # print("encrypted message:", pow(data, e, n))
    # print("decrypted message:", pow(data, d, n))

    # pt = int(input("Enter plain text: "))
    # ct = pow(pt, e, n)
    # print(f"Cipher text = {ct}")

    # dt = pow(ct, d, n)
    # print(f"Decrypted text = {dt}")


if __name__ == '__main__':
    main()
